perspective_reddit_bot/check_mod_actions.py

```diff
@@ -51,66 +51,163 @@
+  """Read lines from file.
+
+  Args:
+    input_path: (str) where to read from
+    stop_at_eof: (bool) returns reaching end of file if true.
+    yield_every_period: (timedelta) yields sentinel value every time period if
+        stream doesn't have contents.
+    sentinel: value yielded every yield_every_period
+
+  Yields: lines from input_path, or sentinel values
+  """
+  with open(input_path) as f:
+    last_time = datetime.utcnow()
+    while True:
+      now = datetime.utcnow()
+      where = f.tell()
+      line = f.readline()
+      if line:
+        print('.', end='')
+        sys.stdout.flush()
+        yield line
+        last_time = now
+      elif stop_at_eof:
+        print('\nread_lines stopping due to stop_at_eof.')
+        return
+      elif now - last_time > yield_every_period:
+        yield sentinel
+        last_time = now
+      else:
+        print('\n(eof, waiting {})'.format(yield_every_period))
+        time.sleep(yield_every_period.total_seconds())
+        f.seek(where)
+
+
+_PRAW_BATCH_SIZE = 100
+
+
+def read_batches(input_path, stop_at_eof, max_batch_delay):
+  """Yields batches from input_path."""
+  current_batch = []
+  last_yield_time = datetime.utcnow()
+  sentinel = object()
+  for line in read_lines(input_path, stop_at_eof, max_batch_delay, sentinel):
+    now = datetime.utcnow()
+    if line is not sentinel:
+      current_batch.append(line)
+    if (current_batch
+        and (len(current_batch) == _PRAW_BATCH_SIZE
+             or now - last_yield_time > max_batch_delay)):
+      print('batch(', len(current_batch), ')')
+      yield current_batch
+      current_batch = []
+      last_yield_time = now
+
+  # read_lines exited
+  if current_batch:
+    yield current_batch
+
+
+def read_records_with_wait(input_path, stop_at_eof, max_batch_delay,
+                           timestamp_key, wait_delta):
+  """Yields batches, waiting until all items are old enough."""
+  for batch in read_batches(input_path, stop_at_eof, max_batch_delay):
+    records = [json.loads(line) for line in batch]
+    youngest = max(r[timestamp_key] for r in records)
+    create_time_utc = datetime.strptime(youngest, '%Y%m%d_%H%M%S')
+    wait_until(create_time_utc + wait_delta)
+    yield records
+
+
+def prefix_comment_id(i):
+  return i if i.startswith('t1_') else 't1_' + i
+
+
+# TODO: this function is complicated because reddit.info() _may_ omit data
+# (i.e. given 100 comments, it may return fewer than 100 Comment objects). so,
+# there's extra bookkeeping to check for dropped records. if this ~never
+# happens, we can simplify this code significantly.
+def check_mod_actions(output_path, reddit, comments, id_key, has_mod_creds):
+  comment_ids = [prefix_comment_id(c[id_key]) for c in comments]
+  statuses = fetch_comment_statuses(reddit, comment_ids, has_mod_creds)
+
+  # Index comments by ID so we can attach the status data to the comment
+  # records. (Cannot comments with statuses, since reddit.info() may drop
```

Comment:
Typo - missing word but I'm not sure which.

Reply:
"zip" :) fixed